### PR TITLE
Trade enhancement - ability to POST new trade, enhance Trade serializer

### PR DESCRIPTION
**Pull request by cristianlucaci**
@marianlungu  pls review, will enhance trading url to allow PUT, to update trade once created.
no need to merge yet.

### CODE REVIEW

**Comment by cristianlucaci** [comment]
@marianlungu pls review again

File: troccproducts/models.py [hunk 1]
```diff
@@ -58,27 +92,100 @@ def get_cover(self):
...
+        Validate instance data, before saving the instance in db.
+        1. TradeFor is required
+        2. If closing a trade, both tradeIn and tradeFor are required.
+        """
+        # Perform the data validation.
+        if self.tradeFor.count() == 0:
+            raise ValidationError({'tradeFor': 'This field is required and can not be empty.'})
+
+        # if trying to close the trade
+        if self.status in (ACCEPTED, CLOSED) and (self.tradeFor.count() == 0 or self.tradeIn.count() == 0):
```
Comment (marianlungu0001) [inline troccproducts/models.py]: @cristianlucaci Isn't the status updated after this method is being called ?. The state will be in OPEN when it enters here. full_clean is called before status_update

File: troccproducts/models.py [hunk 2]
```diff
@@ -58,27 +92,100 @@ def get_cover(self):
...
+            raise ValidationError({'tradeFor': 'This field is required and can not be empty.',
+                                   'tradeIn': 'This field is required and can not be empty.'})
+        super(Trade, self).clean()
+
+    def __update_status(self, old_status, new_status):
+        if Trade.is_state_valid(old_status, new_status):
+            self.status = new_status
+            # deactivate all products from tradeIn/tradeFor
+            if new_status in (ACCEPTED, CLOSED):
+                for product in self.tradeIn.all():
```
Comment (marianlungu0001) [inline troccproducts/models.py]: it's not a big deal :P but I think it looks better to have a status variable and assign the INACTIVE or ACTVE, and after that loop through the products and assign the var

File: troccproducts/models.py [hunk 3]
```diff
@@ -58,27 +92,100 @@ def get_cover(self):
...
+                    product.update_status(Product.INACTIVE)
+                for product in self.tradeFor.all():
+                    product.update_status(Product.INACTIVE)
+            # if trade is not confirmed, reactivate all products
+            if new_status in (INACTIVE,):
+                for product in self.tradeIn.all():
```
Comment (marianlungu0001) [inline troccproducts/models.py]: I think something like this should work. Performance wise is worse but it looks prettier :)

for (tradinProduct, tradeForProduct) in zip(self.tradeIn.all(), self.tradeFor.all()):

File: troccproducts/serializers.py [hunk 4]
```diff
@@ -108,13 +105,105 @@ def _extract_category_object(self, data):
...
 
 
-class CommentSerializer(serializers.ModelSerializer):
-	class Meta:
-		model = Comment
-		fields = ('id', 'user', 'created', 'comment', 'product', 'trade', 'active')
-
-
+# ? change to hyperlink?
 class TradeSerializer(serializers.ModelSerializer):
```
Comment (marianlungu0001) [inline troccproducts/serializers.py]: Awesome work :+1:

File: troccproducts/models.py (hunk 1 above)
Comment (cristianlucaci) [inline troccproducts/models.py]: clean will be called before save, but its used to validate at least 1 product was set for the trade on both In/For sides, if trying to close a trade, its only a validation so makes sense to add it in clean. 
The validation from __update_status could potentially be moved in clean() though.. and only if clean returns true, it continues to save()

File: troccproducts/models.py (hunk 2 above)
Comment (cristianlucaci) [inline troccproducts/models.py]: pls feel free to update it before merge

File: troccproducts/models.py (hunk 3 above)
Comment (cristianlucaci) [inline troccproducts/models.py]: not familiar with the syntax, pls feel free to update it before merge

**Pull request merged**